docs/design/README.md

```diff
@@ -0,0 +1,551 @@
+```
+
+These metadata files will give twoliter the information it needs to resolve, pull, and aggregate all dependency kits needed for a build.
+This dependency resolution may be kept in a lock file inspired by `Cargo.lock` if we find it helpful.
+
+## Containers
+
+In order to minimize the amount of software required on the builder's system, Twoliter will use containers to drive the builds.
+There are two containers that will be published to facilitate the build system.
+
+### Twoliter Container
+
+Twoliter will use this container to bootstrap the build process.
+It needs to contain items the existing Bottlerocket build system expects to be available on the host.
+Fedora will be its base as it is for the SDK.
+
+The image will include the following:
+- the docker CLI
+- cargo make
+- buildsys, pubsys, testsys, etc.
+- Scripts accessed by cargo make, such as `docker-go`.
+- `Makefile.toml`
+
+It is within this container that Twoliter will invoke `cargo make` to initiate a build.
+
+### SDK Changes
+
+The existing build system expects access to the mounted core project git tree for certain things during the build.
+These will not be available and need to be added to the Bottlerocket SDK:
+- Scripts used by `docker build` commands such as `rpm2img`, `rpm2kmodkit` and `partplanner`.
+- RPM macros.
+
+### Docker Domain Socket
+
+Twoliter will mount the host's docker socket (i.e. `/var/run/docker.sock`).
+This means that `docker` commands inside the Twoliter container environment will be interacting with the host's daemon.
+This presents certain challenges because directories need to be mounted from "within" this container.
+In particular:
+
+* The Twoliter container process needs to have `docker` group permissions.
+* The Twoliter container process needs to have permissions to write in certain host directories.
+
+We can solve this by introspecting the Twoliter CLI's process to get the UID and GID it is invoked with.
+We make the assumption that the user invoking the CLI has the permissions to interact with docker and to write in the desired host directories.
+We then pass `--user "uid:gid"` to docker when starting the Twoliter container.
+We can also allow the Twoliter command line to accept uid and gid arguments to override this behavior.
+
+An additional problem to solve is that the paths to mounted directories need to be absolutely the same on the host as they are in the Twoliter container.
+This will not be problem unless the host build or sourcecode directory is in some wildly unusual place, like `/etc/docker` or `/usr/bin`.
+We can prevent this by making a certain set of top-level directories off-limits on the host for the directories we need to mount.
+An error message will describe the problem instead of running the container and failing due to occluded directories.
+
+## Twoliter Projects
+
+Twoliter will have the concept of a project.
+A project can contain multiple variant definitions, multiple kit definitions, and multiple packages.
```

Comment:
One possible use-case is that we could move all `aws-k8s-*` variants out-of-tree to the same twoliter project, thereby sharing the code they can, apart from kubelet builds.

Another example could be that I'm publishing a Bottlerocket appliance and I need a build which works on AWS and another which works on e.g. Raspberry Pi. The installed software would largely intersect, so it would be convenient to define them in the same source tree.